graph/appender/security_policy.go

```diff
@@ -0,0 +1,145 @@
+// The appender currently reports only mutual_tls security although is written in a generic way.
+type SecurityPolicyAppender struct {
+	Duration     time.Duration
+	GraphType    string
+	IncludeIstio bool
+	QueryTime    int64 // unix time in seconds
+}
+
+// AppendGraph implements Appender
+func (a SecurityPolicyAppender) AppendGraph(trafficMap graph.TrafficMap, namespace string) {
+	if len(trafficMap) == 0 {
+		return
+	}
+
+	client, err := prometheus.NewClient()
+	checkError(err)
+
+	a.appendGraph(trafficMap, namespace, client)
+}
+
+func (a SecurityPolicyAppender) appendGraph(trafficMap graph.TrafficMap, namespace string, client *prometheus.Client) {
+	// query prometheus for mutual_tls info in two queries:
+	// 1) query for active security originating from a workload outside the namespace
+	groupBy := "source_workload_namespace,source_workload,source_app,source_version,destination_service_namespace,destination_service_name,destination_workload,destination_app,destination_version,connection_security_policy"
+	query := fmt.Sprintf("sum(rate(%s{reporter=\"destination\",source_workload_namespace!=\"%v\",destination_service_namespace=\"%v\",connection_security_policy!=\"none\",response_code=~\"%v\"}[%vs])) by (%s)",
+		"istio_requests_total",
+		namespace,
+		namespace,
+		"[2345][0-9][0-9]",        // regex for valid response_codes
+		int(a.Duration.Seconds()), // range duration for the query
+		groupBy)
+	outVector := promQuery(query, time.Unix(a.QueryTime, 0), client.API())
+
+	// 2) query for active_security originating from a workload inside of the namespace
+	istioCondition := ""
+	if !a.IncludeIstio {
+		istioCondition = fmt.Sprintf(",destination_service_namespace!=\"%s\"", config.Get().IstioNamespace)
+	}
+	query = fmt.Sprintf("sum(rate(%s{reporter=\"destination\",source_workload_namespace=\"%v\"%s,connection_security_policy!=\"none\",response_code=~\"%v\"}[%vs])) by (%s)",
+		"istio_requests_total",
+		namespace,
+		istioCondition,
+		"[2345][0-9][0-9]",        // regex for valid response_codes
+		int(a.Duration.Seconds()), // range duration for the query
+		groupBy)
+	inVector := promQuery(query, time.Unix(a.QueryTime, 0), client.API())
+
+	// create map to quickly look up responseTime
+	securityPolicyMap := make(map[string]string)
+	a.populateSecurityPolicyMap(securityPolicyMap, &outVector)
+	a.populateSecurityPolicyMap(securityPolicyMap, &inVector)
+
+	applySecurityPolicy(trafficMap, securityPolicyMap)
+}
+
+func applySecurityPolicy(trafficMap graph.TrafficMap, securityPolicyMap map[string]string) {
+	for _, s := range trafficMap {
+		for _, e := range s.Edges {
+			key := fmt.Sprintf("%s %s", e.Source.ID, e.Dest.ID)
+			if securityPolicy, ok := securityPolicyMap[key]; ok {
+				switch securityPolicy {
+				case "mutual_tls":
+					e.Metadata["isMTLS"] = true
+				default:
+					log.Debugf("Skipping unhandled security policy [%s]", securityPolicy)
```

Comment:
will this spam the debug logs when no security is enabled? Because I believe the value will be "none" but you don't have a case "none" here, so the default will execute and thus you will log "skipping unhandled security policy" everytime "none" is found.

Reply:
The query specifies connection_security_policy!="none".  So that should not be an issue.